File: ProgrammingChallengesAndOnlineJudges/Codechef/FEB12/WCOUNT/wcount2.c
#include <stdio.h>
#include <string.h>

#define MOD 1000000007

typedef unsigned long long int ulli;

char s[501];

ulli factorial(ulli n)
{
	if(!n)
		return 1;
	ulli f = 1, i = 1;
	while(i <= n)
		f *= i++;
	return f;
}

int main(void)
{
	int t, nr, dr[52], i, j, terms;
	char c;
	ulli result, temp;
	scanf("%d", &t);
	while(t--)
	{
		scanf("%s", s);
		nr = strlen(s);
		
		memset(dr, 0, 52 * sizeof(int));
		for(i = 0; c = s[i]; i++)
			(c >= 'a' && c <= 'z') ? ++dr[c - 'a'] : ++dr[c - 'A' + 26];

		for(i = 0, result = 1, terms = 0; i < 52; terms += dr[i++])
		{
			for(j = terms + 1, temp = 1; j <= terms + dr[i]; j++)
				temp *= j;
			result = (result % MOD) * (temp / factorial(dr[i]));
		}

		for(i = terms + 1; i <= nr; i++)
			result = (result % MOD) * i;

		printf("%d\n", (int) (result % MOD));
	}
	return 0;
}

File: ProgrammingChallengesAndOnlineJudges/Codechef/FEB12/WCOUNT/wcount2.py
#!/usr/bin/python

import math
import sys
import string

mod = 1000000007
t = raw_input()
for s in sys.stdin:
	s = s.strip()
	nr = len(s)
	if nr == 0:
		continue
	dr = []
	while len(s) > 0:
		c = s[0]
		dr.append(s.count(c))
		s = s.replace(c, '')
	result = 1
	terms = 0
	for i in dr:
		temp = 1
		for j in range(terms + 1, terms + i + 1):
			temp *= j
		result = (result % mod) * (temp / math.factorial(i))
		terms += i
	for i in range(terms + 1, nr + 1):
		result = (result % mod) * i
	print result % mod
sys.exit(0)
